zebra-state/src/service/finalized_state/disk_format/upgrade.rs

```diff
@@ -0,0 +1,511 @@
+                // We do this on a best-effort basis for versions that are still supported.
+            }
+        }
+    }
+
+    /// Apply any required format updates to the database.
+    /// Format changes should be launched in an independent `std::thread`.
+    ///
+    /// If `cancel_receiver` gets a message, or its sender is dropped,
+    /// the format change stops running early.
+    //
+    // New format upgrades must be added to the *end* of this method.
+    fn apply_format_upgrade(
+        self,
+        config: Config,
+        network: Network,
+        initial_tip_height: Option<Height>,
+        cancel_receiver: mpsc::Receiver<CancelFormatChange>,
+    ) {
+        let Upgrade {
+            newer_running_version,
+            older_disk_version,
+        } = self
+        else {
+            unreachable!("already checked for Upgrade")
+        };
+
+        // # New Upgrades Sometimes Go Here
+        //
+        // If the format change is outside RocksDb, put new code above this comment!
+        let Some(initial_tip_height) = initial_tip_height else {
+            // If the database is empty, then the RocksDb format doesn't need any changes.
+            info!(
+                ?newer_running_version,
+                ?older_disk_version,
+                "marking empty database as upgraded"
+            );
+
+            Self::mark_as_upgraded_to(&database_format_version_in_code(), &config, network);
+
+            info!(
+                ?initial_tip_height,
+                ?newer_running_version,
+                ?older_disk_version,
+                "database is fully upgraded"
+            );
+
+            return;
+        };
+
+        // Example format change.
+        //
+        // TODO: link to format upgrade instructions doc here
+
+        // Check if we need to do this upgrade.
+        let database_format_add_format_change_task =
+            Version::parse("25.0.2").expect("version constant is valid");
+
+        if older_disk_version < database_format_add_format_change_task {
+            let mut upgrade_height = Height(0);
+
+            // Go through every height from genesis to the tip of the old version.
+            // If the state was downgraded, some heights might already be upgraded.
+            // (Since the upgraded format is added to the tip, the database can switch between
+            // lower and higher versions at any block.)
+            //
+            // Keep upgrading until the initial database has been upgraded,
+            // or this task is cancelled by a shutdown.
+            while upgrade_height <= initial_tip_height
+                && matches!(cancel_receiver.try_recv(), Err(mpsc::TryRecvError::Empty))
+            {
+                // TODO: Do one format upgrade step here
+
+                upgrade_height = (upgrade_height + 1).expect("task exits before maximum height");
+            }
```

Comment:
There's a lot of task code, versioning code, and test code in this PR. But the main part of it is here - this is where we can add the code to do the format upgrades.

We can call whatever database methods we like here, and they can take as long as is needed, because this code runs in a separate thread.

I just noticed I need to pass the database into this code, then we can call it! But that's an easy change, the database can be cloned here:
https://github.com/ZcashFoundation/zebra/pull/7031/files#diff-17dd8a7e992ef6adf7d3c5e78546718069747564f3b60ab3662747a0ceaddc68R84

Reply:
The database is available in this code now.